### PR TITLE
New methods: BaseActiveRecord::loadRelations() and BaseActiveRecord::loadRelationsFor().

### PR DESCRIPTION
**Pull request by PowerGamer1**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #11000, #12743

Equivalent of https://laravel.com/docs/10.x/eloquent-relationships#lazy-eager-loading in Eloquent ORM.

Should be mentioned in Yii2 guide.

### CODE REVIEW

File: framework/db/BaseActiveRecord.php
Comment (samdark) [inline framework/db/BaseActiveRecord.php:1793]:
```suggestion
     * if (Yii:app()->getUser()->getIdentity()->canAccessOrders()) {
```

File: framework/db/BaseActiveRecord.php
Comment (samdark) [inline framework/db/BaseActiveRecord.php:1808]:
```suggestion
        if (empty($models)) {
```

**Comment by samdark** [comment]
Would you please add tests for the method?

**Comment by PowerGamer1** [comment]
> Would you please add tests for the method?

Done.

File: framework/db/BaseActiveRecord.php
Comment (mtangoo) [inline framework/db/BaseActiveRecord.php:1808]: Why not ` if (!empty($models))` and avoid empty `return` statement which raises Codecov warning?

Comment (PowerGamer1) [reply framework/db/BaseActiveRecord.php:1808]: It will make the code less readable IMO:
- your variant uses extra logical operator (which makes expression more complex for human understanding);
- the ! operator is easily missable at a glance;
- my variant has a clearly visible useful work of a function (last line) and a clearly separated condition when it can be skipped; your variant would have these tangled up making it harder to understand what is the normal function flow and what are the exceptions in it.

Making the code less readable just to shut up a warning in some imperfect external tool is a bad idea.

Comment (terabytesoftw) [reply framework/db/BaseActiveRecord.php:1808]: But shouldn't it cover the test, for when the model is empty?

Comment (PowerGamer1) [reply framework/db/BaseActiveRecord.php:1808]: > But shouldn't it cover the test, for when the model is empty?

You mean if there should be a test for a `loadRelationsFor([], /*...*/)` call? I don't think this test will be very useful.

Comment (terabytesoftw) [reply framework/db/BaseActiveRecord.php:1808]: But it is testing the code, in the case that the model is empty.

Comment (PowerGamer1) [reply framework/db/BaseActiveRecord.php:1808]: > But it is testing the code, in the case that the model is empty.

So you just want to do the testing for the sake of doing testing even if it doesn't test anything useful at all? If you disagree, please show us how such test could possibly look and explain its usefulness.

Comment (mtangoo) [reply framework/db/BaseActiveRecord.php:1808]: @PowerGamer1 when things breaks for empty models with some bug in the code (let say someone else modified your now empty method to something else, how will we detect it? That's where this test comes to rescue. Currently sound like useless, but it will make code "future proof" sort of!

Comment (PowerGamer1) [reply framework/db/BaseActiveRecord.php:1808]: > @PowerGamer1 ... let say someone else modified your now empty return to something else, how will we detect it?

At present, testing a single return statement is useless. If, in the future, as you say someone replaces/modifies this return statement with a MORE COMPLEX piece of code - then at that point in time testing such MORE COMPLEX piece of code might become warranted. Obviously, testing of such (currently non-existent) more complex code path should be done when it is introduced by the person who introduced it.

Also, if I were you (i.e. engine maintainer), I would be much more concerned with testing infinitely more complex and currently completely untested pieces of code (such as `ActiveQueryTrait::findWith()` and myriad others). In fact, the only reason I am currently writing any tests at all is the fact that new shortcut methods rely on such untested functionality - i.e. I don't really test NEW code, I have to test already EXISTING code.

P.S. That test you want me to write - have you even imagined how exactly it will look? What exactly are you going to assert (like `assertTrue()` etc.) in it?

**Comment by rhertogh** [comment]
It looks to me that the code relies on the "inverseOf" definition of the related ActiveRecord hasOne/hasMany relation.
Is that correct? And if so, what happens when there is no "inverseOf" relation defined?

**Comment by PowerGamer1** [comment]
> It looks to me that the code relies on the "inverseOf" definition of the related ActiveRecord hasOne/hasMany relation. Is that correct? And if so, what happens when there is no "inverseOf" relation defined?

No such reliance on the `inverseOf` at all. The new methods just call (in a convenient way) the same existing functionality that is executed when you perform the eager loading using `with()` in a call like:
```PHP
Customer::find()->with('orders.items')->all();
```
That example above works without defining inverse relation and so do new methods which you can see in the tests (neither `Customer::getOrders()` nor `Order::getItems()` relation in the tests uses `inverseOf`).

**Comment by rhertogh** [comment]
@PowerGamer1 Yes, when I dived deeper into it I came to the same conclusion and removed my comment. But your response beat me to it 😄.

**Review by samdark — APPROVED** [review]

**Comment by PowerGamer1** [comment]
> @PowerGamer1 All good. Would you please resolve conflicts so I can merge it?

Done.

**Comment by samdark** [comment]
Please resolve the conflicts so I can hit the merge button. Thanks.

**Comment by samdark** [comment]
Please resolve the conflicts so I can hit the merge button. Thanks.

**Comment by samdark** [comment]
Damn :( Conflicts again. @PowerGamer1 need your help.

**Comment by PowerGamer1** [comment]
You might want to merge this soon as I will not be able to rebase this commit to latest master for much longer (as I will be leaving github due to forced 2FA policy).

**Comment by bizley** [comment]
@PowerGamer1 could you resolve conflicts one more time? I'll merge this immediately after. Sorry for the long time waiting.

**Comment by PowerGamer1** [comment]
> @PowerGamer1 could you resolve conflicts one more time? I'll merge this immediately after. Sorry for the long time waiting.

@bizley No worries, conflicts resolved.

**Comment by bizley** [comment]
Thank you. Sorry to hear you are leaving GH.

**Comment by xicond** [comment]
I'm wandering, is this support nested relationship like dot notation

**Pull request merged**
